utils/sharding.go

```diff
@@ -0,0 +1,212 @@
+func (r *ShardHelper) ShouldOwn(input string) bool {
+	return r.GetShardOwner(input) == r.PodIP
+}
+
+// SetMembers accepts a map where the keys are member IPs and uses those IPs as the members for sharding.
+func (r *ShardHelper) SetMembers(newMembers map[string]struct{}) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// Add new members. Add returns immediately if the member already exists.
+	for newMember := range newMembers {
+		r.ring.Add(peer(newMember))
+	}
+
+	// Remove members which don't exist anymore.
+	for _, oldMember := range r.ring.GetMembers() {
+		if _, ok := newMembers[oldMember.String()]; !ok {
+			r.ring.Remove(oldMember.String())
+		}
+	}
+}
+
+// SetMembersFromLists is a wrapper around SetMember which accepts slices instead of a map.
+func (r *ShardHelper) SetMembersFromLists(lists ...[]string) {
+	members := make(map[string]struct{})
+	for _, l := range lists {
+		for _, m := range l {
+			members[m] = struct{}{}
+		}
+	}
+	r.SetMembers(members)
+}
+
+// Helper type for members of peer ring.
+type peer string
+
+func (p peer) String() string {
+	return string(p)
+}
+
+func BuildPeerHashRing(consistentCfg consistent.Config, podIP string, serviceName string, serviceNamespace string) *ShardHelper {
+	consistentHashRing := consistent.New(nil, consistentCfg)
+	mutex := sync.RWMutex{}
+	return &ShardHelper{
+		PodIP:            podIP,
+		ServiceName:      serviceName,
+		ServiceNamespace: serviceNamespace,
+		mu:               &mutex,
+		ring:             consistentHashRing,
+	}
+}
+
+func BuildPeerInformer(stopper chan struct{}, peerRing *ShardHelper, ringConfig consistent.Config, log logr.Logger) cache.SharedIndexInformer {
+
+	dc, err := dynamic.NewForConfig(ctrl.GetConfigOrDie())
+	if err != nil {
+		log.Error(err, "unable to set up informer")
+		os.Exit(1)
+	}
+
+	listOptionsFunc := dynamicinformer.TweakListOptionsFunc(func(options *metav1.ListOptions) {
+		options.FieldSelector = "metadata.name=" + peerRing.ServiceName
+	})
+
+	// Use our namespace and expected endpoints name in our future informer.
+	factory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(dc, 0, peerRing.ServiceNamespace, listOptionsFunc)
+
+	// Construct an informer for the endpoints.
+	informer := factory.ForResource(schema.GroupVersionResource{
+		Group: "", Version: "v1", Resource: "endpoints"}).Informer()
+
+	// Set handlers for new/updated endpoints.
+	handlers := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			added, kept, _, ok := getEndpointChanges(obj, nil, log)
+			if !ok {
+				return
+			}
+			peerRing.SetMembersFromLists(added, kept)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			// In the future, we might need to re-queue objects which belong to deleted peers.
+			// When scaling down, it is possible that metrics will be double reported for up to the reconciliation period.
+			// For now, we'll just set the desired peers.
+			added, kept, _, ok := getEndpointChanges(newObj, oldObj, log)
```

Comment:
This is not perfect. For fast changes we might get an outdated object in `newObj` from the cache. We should always check against the k8s API if the `newObj` resourceversion is still the latest.

Reply:
WDYT of asking for review from sig-operator? I just see it as a concern but am not certain how valid that concern is :D

Reply:
That's a good idea, I've requested a review from the sig